server/posts/queries.ts

```diff
@@ -306,83 +182,118 @@ export const postNewContribution = async (
+      `Either parentPostId (${parentPostId}) or threadExternalId (${threadExternalId}) is 
+       required when creating a new contribution.`
     );
-    let {
-      board_slug,
-      board_external_id,
-      user_id,
-      username,
-      user_avatar,
-      secret_identity_name,
-      secret_identity_avatar,
-      secret_identity_color,
-      accessory_avatar,
-      thread_id,
-      thread_external_id,
-      post_id = null,
-    } = await getThreadDetails(t, {
-      identityId,
-      accessoryId,
-      parentPostId,
-      threadExternalId,
-      firebaseId,
-    });
-    const result = await t.one(sql.makePost, {
-      post_string_id: uuidv4(),
-      parent_post: post_id,
-      parent_thread: thread_id,
-      user_id,
+
+    let threadData = await transaction.one(
+      parentPostId ? sql.getPostDetails : sql.getThreadDetails,
+      {
+        post_string_id: parentPostId,
+        thread_external_id: threadExternalId,
+        parent_comment_string_id: null,
+        firebase_id: firebaseId,
+      }
+    );
+
+    if (!threadData.secret_identity_id && !threadData.role_identity_id) {
+      // User has no identity assigned within this thread, so we assign it.
+      const threadIdentityDetails = await addNewIdentityToThread(transaction, {
+        // If an identityId was passed, then we'll try to add that specific identity
+        // if not, we'll let addNewIdentityToThreadByBoardId decide how to deal with
+        // assigning a new identity.
+        identityExternalId: contributionData.identityId ?? null,
+        accessory_id: contributionData.accessoryId ?? null,
+        thread_id: threadData.thread_id,
+        firebaseId,
+        boardExternalId: threadData.board_external_id,
+      });
+
+      // We update thread data with the newly returned identity details
+      threadData = {
+        ...threadData,
+        ...threadIdentityDetails,
+      };
+    }
+
+    const {
       content,
+      anonymityType,
+      whisperTags,
+      indexTags,
+      contentWarnings,
+      categoryTags,
+    } = contributionData;
+    const postedContribution = await transaction.one(sql.makePost, {
+      post_string_id: uuidv4(),
+      parent_post: threadData.post_id ?? null,
+      parent_thread: threadData.thread_id,
+      // TODO: we should remove references to firebaseId from within the server and just make it
+      // so the internal user_id is what the server always deals with.
+      // This way we won't have to take this information from threadData, where it really doesn't
+      // belong.
+      user_id: threadData.user_id,
+      content: content,
       anonymity_type: anonymityType,
       whisper_tags: whisperTags,
     });
-    log(`Added new contribution to thread ${thread_id}.`);
-    log(result);
+    log(`Added new contribution to thread ${threadData.thread_id}.`);
+    log(postedContribution);
```

Comment:
Just wanted to thank you for this change lmao

Reply:
1. Write a terribly messy codebase
2. Fix the mess you made yourself
3. ???
4. Profit